### PR TITLE
Satellite Self-Observability

### PR DESCRIPTION
**Pull request by EvanLjp**
add prometheus telemetry

### CODE REVIEW

File: dist/licenses/LICENSE-prometheus-client-golang.txt
Comment (wu-sheng) [inline dist/licenses/LICENSE-prometheus-client-golang.txt:1]: Apache 2.0 license is not required to be copied. If NOTICE exists, you need to copy the content into `dist/NOTICE`.

File: dist/LICENSE
Comment (wu-sheng) [inline dist/LICENSE:216]: Note this sentence.

File: configs/satellite_config.yaml
Comment (wu-sheng) [inline configs/satellite_config.yaml:37]: What does `pipes` mean?

File: plugins/server/prometheus/Prometheus_test.go
Comment (wu-sheng) [inline plugins/server/prometheus/Prometheus_test.go:56]: You have only tested the Prometheus endpoint is open, right?

File: internal/pkg/plugin/definition.go
Comment (surlymo) [inline internal/pkg/plugin/definition.go:60]: magic num

File: internal/pkg/plugin/definition.go
Comment (surlymo) [inline internal/pkg/plugin/definition.go:54]: why not just return the string name?

File: plugins/fallbacker/timer/timer_fallbacker.go
Comment (surlymo) [inline plugins/fallbacker/timer/timer_fallbacker.go:31]: why all need this struct

File: internal/pkg/plugin/registry.go
Comment (kezhenxu94) [inline internal/pkg/plugin/registry.go:107]:
```suggestion
		panic(fmt.Errorf("%s plugin must have a field named CommonField.", plugin.Name()))
```

File: internal/pkg/plugin/registry.go
Comment (kezhenxu94) [inline internal/pkg/plugin/registry.go:52]: I didn't notice this before, but `fmt.Printf` doesn't allow users to specify verbosity of logs, many other places are using `fmt.Printf` too

File: internal/pkg/plugin/registry.go
Comment (kezhenxu94) [inline internal/pkg/plugin/registry.go:110]: Use constant `TagName`

File: tools/install_protoc.sh
Comment (kezhenxu94) [inline tools/install_protoc.sh:21]: I saw there is a binary for MacOS too, why not support MacOS here as well?

Comment (wu-sheng) [reply tools/install_protoc.sh:21]: I think he is setting CI? Which only include Ubuntu for now? We should set up the MacOS for CI testing too.

File: internal/satellite/telemetry/telemetry.go
Comment (hanahmily) [inline internal/satellite/telemetry/telemetry.go:35]: IIRC, we should not introduce Kubernetes notions into satellite since it's designed to deployed on VM as well. 

How about using more general and SkyWalking-style concepts, for example, `Cluster`, `Service` and `Instance`

File: plugins/server/prometheus/Prometheus.go
Comment (hanahmily) [inline plugins/server/prometheus/Prometheus.go:1]: Why the file name is a capical letter?

File: plugins/server/prometheus/Prometheus_test.go
Comment (hanahmily) [inline plugins/server/prometheus/Prometheus_test.go:33]: Could you use different endpoints and ports instead of defaults.

File: .github/workflows/build-and-test.yaml
Comment (hanahmily) [inline .github/workflows/build-and-test.yaml:40]: @kezhenxu94 Could we use a third part plugin now?

File: tools/install_protoc.sh
Comment (hanahmily) [reply tools/install_protoc.sh:21]: In this PR, he's using a third-party plugin instead of this bash to install protoc

**Review by hanahmily — COMMENTED** [review]
Some nits

File: .github/workflows/build-and-test.yaml
Comment (kezhenxu94) [reply .github/workflows/build-and-test.yaml:40]: > @kezhenxu94 Could we use a third part plugin now?

This kind of actions are blocked, unless we get enough reasons to persuade the infra team to unblock it.

File: tools/install_protoc.sh
Comment (kezhenxu94) [reply tools/install_protoc.sh:21]: > I think he is setting CI? Which only include Ubuntu for now? We should set up the MacOS for CI testing too.

Locally, when you run `make deps` to setup the dev env, this is executed and if you didn't install protoc correctly, it fails.

**Comment by wu-sheng** [comment]
> In this PR, he's using a third-party plugin instead of this bash to install protoc

I think 3rd party github action has been blocked by INFRA. @hanahmily

File: internal/satellite/module/gatherer/receiver_gatherer.go
Comment (gxthrj) [inline internal/satellite/module/gatherer/receiver_gatherer.go:59]: I am not very clear about this. Should the metrics here be defined in plugins ? Or we supports prometheus in `Gather` by default?

File: configs/satellite_config.yaml
Comment (EvanLjp) [reply configs/satellite_config.yaml:37]: the namespace concept is similar to the Kubernetes namespaces, so I change the original namespaces concept to the new name call pipe. Each pipe has one gatherer, one processor, and one sender.

File: plugins/server/prometheus/Prometheus_test.go
Comment (EvanLjp) [reply plugins/server/prometheus/Prometheus_test.go:56]: yes

File: internal/pkg/plugin/definition.go
Comment (EvanLjp) [reply internal/pkg/plugin/definition.go:54]: Use the directory structure to ensure that the plug-in name is unique

File: plugins/fallbacker/timer/timer_fallbacker.go
Comment (EvanLjp) [reply plugins/fallbacker/timer/timer_fallbacker.go:31]: the common field is inject by boot process.
for example:  a plugins want to know which pipe belong to, the commonField is for this.

File: internal/satellite/telemetry/telemetry.go
Comment (EvanLjp) [reply internal/satellite/telemetry/telemetry.go:35]: no problem

File: internal/satellite/module/gatherer/receiver_gatherer.go
Comment (EvanLjp) [reply internal/satellite/module/gatherer/receiver_gatherer.go:59]: the feature is to monitor self status.  Could you explain that again? I'm not very clear

File: plugins/server/prometheus/Prometheus_test.go
Comment (wu-sheng) [reply plugins/server/prometheus/Prometheus_test.go:56]: Got it.

File: configs/satellite_config.yaml
Comment (wu-sheng) [reply configs/satellite_config.yaml:37]: Do you update the doc? I think the `namespace` has been mentioned there.

Comment (EvanLjp) [reply configs/satellite_config.yaml:37]: sorry, thx for your notice

File: plugins/server/prometheus/Prometheus_test.go
Comment (EvanLjp) [reply plugins/server/prometheus/Prometheus_test.go:33]: image
could set custom values

Comment (EvanLjp) [reply plugins/server/prometheus/Prometheus_test.go:33]: image

File: .github/workflows/build-and-test.yaml
Comment (EvanLjp) [reply .github/workflows/build-and-test.yaml:40]: image
@hanahmily  i use shell to install protoc

**Comment by EvanLjp** [comment]
@gxthrj @nic-chen  @kezhenxu94  how about to use init() in logger? can't use init() in Logger brings up a lot of chore things.

File: internal/satellite/module/gatherer/receiver_gatherer.go
Comment (gxthrj) [reply internal/satellite/module/gatherer/receiver_gatherer.go:59]: > the feature is to monitor self status

OK, This is the metric for satellite self.
I mistakenly thought that this is an extension of Prometheus Gather.

Comment (wu-sheng) [reply internal/satellite/module/gatherer/receiver_gatherer.go:59]: @EvanLjp I think you should rename this PR to `Satellite Self-Observability` :) I used to think this is for scratch from prometheus endpoint.

File: tools/install_protoc.sh
Comment (wu-sheng) [reply tools/install_protoc.sh:21]: I tried, I have to install the protoc manually.

**Review by wu-sheng — APPROVED** [review]
Generally good to me.

**Review by gxthrj — APPROVED** [review]
LGTM

**Review by surlymo — APPROVED** [review]
LGTM

**Pull request merged**
